### ISSUE TITLE
Suggestion to Reduce Memory Usage During Inference

### ISSUE THREAD

**Issue by imczxx** [QwenLM/Qwen2-VL#180]
# Problem Description
In the Prefill stage (i.e., when outputting the first token), calculating logits for all token positions causes significant memory waste. With a vocabulary size of 152,064, the logits for one token position require 152,064 * 4 (float) / 1024^2 = 0.58MB. For input sequences thousands of tokens long, this results in considerable memory waste. This is the part of the model inference process that uses the most memory.

In reality, only the logits for the last token are necessary for inference. Modifying this can significantly reduce memory usage.

# Reproduction Steps
- ⚠️NOTE:  Using Flash Attention 2 and the script from the "Using 🤗 Transformers to Chat" section in the README.

Modify lines 1619-1621 in transformers/models/qwen2_vl/modeling_qwen2_vl.py, calculate memory usage before and after, and print the shape of hidden_states.

<img width="1146" alt="image" src="https://github.com/user-attachments/assets/add4b87e-f23b-4ccb-9c90-7dab84c6e22d">

Here's the output from the program when generating the first token:

<img width="374" alt="image" src="https://github.com/user-attachments/assets/f47d4db6-fe4a-4088-988b-7a612a845c92">

# Optimization Proposal
For samples without labels, keep only the hidden_states of the last token position.

<img width="1131" alt="image" src="https://github.com/user-attachments/assets/17731c20-943d-452d-baf1-f7966422ed30">

# Performance Comparison
## Before
<img width="374" alt="image" src="https://github.com/user-attachments/assets/f47d4db6-fe4a-4088-988b-7a612a845c92">

<img width="1459" alt="image" src="https://github.com/user-attachments/assets/254ef61f-bb00-48eb-85f9-f1ed1b27f804">

## After
<img width="379" alt="image" src="https://github.com/user-attachments/assets/5015c5dd-117d-45dc-8212-2023c4ab6060">

<img width="1463" alt="image" src="https://github.com/user-attachments/assets/ed2f3656-11ba-4177-8db3-5829f37f44ba">

Making this modification doesn't affect the inference process and can significantly reduce memory usage.

**Comment by jklj077** [comment]
Hi, [a similar optimization proposal](https://www.github.com/huggingface/transformers/pull/31292) for memory savings in inference using `num_logits_to_keep` has landed in transformers (but not for newer models). I believe it's a simple change and contributions are welcomed. Do you mind creating a similar PR there?
